audio-echo/app/src/main/java/com/google/sample/echo/PermissionRequestFragment.java

```diff
@@ -0,0 +1,167 @@
+    public void onAttach(Context context) {
+        super.onAttach(context);
+        theInstance = this;
+        callbackPtr = 0L;
+    }
+
+    /**
+     * Static method callable from native code.  This method checks the permission, and then
+     * invokes the callback with the results.  This callback may happen immediately if the permission
+     * state is already known, or it may need to go through the requesting permission and onRequestPermissionsResult
+     * before the callback is called.
+     * <p>If there is already a pending permission check, the callback is invoked immediately  with
+     * and error.</p>
+     * <p>This is marked with API 23 since the permissions are new. The ContextCompat and activity
+     * are aware of the API level and perform the correct behavior for SDKs < 23</p>
+     *
+     * @param permission  - the permission to check.
+     * @param callbackPtr - the pointer to the callback function which has the signature (jint), the
+     *                    value is either Packagemanager.PERMISSION_GRANTED (0),
+     *                    Packagemanager.PERMISSION_DENIED (-1), or another negative number indicating
+     *                    an error state.
+     */
+    @TargetApi(23)
+    public static void checkPermission(final String permission, String rationale, long callbackPtr) {
+        Log.d(TAG,"Checking permission for: " + permission);
+        if (theInstance == null) {
+            Log.e(TAG, "An instance of this fragment has not been attached to the activity.");
+
+            handlePermissionResult(Error_FragmentNotAttached, callbackPtr);
+            return;
+        }
+        if (theInstance.callbackPtr != 0L) {
+            Log.e(TAG, "Permission check already in progress");
+            handlePermissionResult(Error_Busy, callbackPtr);
+            return;
+        }
+
+        theInstance.callbackPtr = callbackPtr;
+
+        int permissionCheck = ContextCompat.checkSelfPermission(theInstance.getActivity(), permission);
+        if (permissionCheck == PackageManager.PERMISSION_DENIED) {
+            Log.d(TAG,"self permission is denied, requesting");
+            // request runtime permission
+            if (ActivityCompat.shouldShowRequestPermissionRationale(theInstance.getActivity(),
+                    permission)) {
+
+                final Activity thisActivity = theInstance.getActivity();
+                View view = thisActivity.getCurrentFocus();
+                if (view == null) {
+                    view = thisActivity.getWindow().getDecorView();
+                }
+                Snackbar.make(view,
+                        rationale,
+                        Snackbar.LENGTH_INDEFINITE).setAction("OK", new View.OnClickListener() {
+                    @Override
+                    public void onClick(View view1) {
+                        theInstance.requestPermissions(new String[]{permission}, RC_REQUEST_PERMISSION);
+                    }
+                }).show();
```

Comment:
I found if I denied it, it crashes and I restarted it, it never asked me to re-try again. I guess this is the code causing the behaviour:
once user denied, user will not be asked again anymore.  She/He has to remove the application cache or remove app totally from the phone/tablet, re-install a new app, only then user could get a chance to re-asked for the permission.

I assume the samples are downloaded, built and run via android studio; user will touch "deny" to see the behaviour,  then user will find out that re-install fact, may be a little un-necessary step for development process

For formal release application, user will have to remove app cache or re-download from play store. 

is this intentional?